ios/brave-ios/Sources/Onboarding/WelcomeFocus/FocusSliderContentView.swift

```diff
@@ -0,0 +1,265 @@
+  @State private var hapticsPlayer: CHHapticPatternPlayer?
+  @State private var hapticsLevel: HapticsLevel = .high
+
+  var leading: Leading
+  var trailing: Trailing
+
+  init(
+    progress: Binding<CGFloat>,
+    @ViewBuilder leading: () -> Leading,
+    @ViewBuilder trailing: () -> Trailing
+  ) {
+    self._progress = progress
+    self.leading = leading()
+    self.trailing = trailing()
+  }
+
+  var isRTL: Bool {
+    layoutDirection == .rightToLeft
+  }
+
+  @ViewBuilder private var actualLeading: some View {
+    if isRTL {
+      trailing
+    } else {
+      leading
+    }
+  }
+
+  @ViewBuilder private var actualTrailing: some View {
+    if isRTL {
+      leading
+    } else {
+      trailing
+    }
+  }
+
+  var body: some View {
+    ZStack {
+      actualLeading
+      actualTrailing
+        .mask {
+          GeometryReader { proxy in
+            Color.black
+              .frame(width: proxy.size.width * (isRTL ? progress : (1 - progress)))
+              .frame(
+                maxWidth: .infinity,
+                maxHeight: .infinity,
+                alignment: isRTL ? .leading : .trailing
+              )
+          }
+        }
+    }
+    .onAppear(perform: prepareSliderHaptics)
+    .onChange(of: progress) { newValue in
+      hapticsLevel = determineHapticIntensityLevel(from: newValue)
+    }
+    .onChange(of: hapticsLevel) { newValue in
+      createContinousHapticFeedback(intensity: newValue)
+    }
+    .overlay {
+      // Fake "grabber"/splitter
+      GeometryReader { proxy in
+        Color(braveSystemName: .textPrimary)
+          .frame(width: 4)
+          .shadow(radius: 8)
+          .overlay {
+            Image("focus-slider-knob", bundle: .module)
+              .padding(.leading, 6)
+          }
+          .offset(x: (proxy.size.width * progress))
+      }
+    }
+    .overlay {
+      // Drag gesture view
+      GeometryReader { proxy in
+        Color.clear
+          .contentShape(Rectangle())
+          .gesture(
+            DragGesture()
+              .updating(
+                $initialProgress,
+                body: { value, state, _ in
+                  // Sets initialProgress when the gesture is active
+                  if state == nil {
+                    state = progress
+                  }
+                }
+              )
+              .onChanged { value in
+                guard let initialProgress else { return }
+                withAnimation(.interactiveSpring) {
+                  progress = max(
+                    0,
+                    min(1, initialProgress + (value.translation.width) / proxy.size.width)
+                  )
+                  if isRTL {
+                    progress = 1 - progress
+                  }
+                }
+              }
+          )
+      }
+    }
+    .padding()
+  }
+
+  private func prepareSliderHaptics() {
+    // Check Hardware is capable
+    guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else { return }
```

Comment:
Is this slow like voice-search hardware checks?

Reply:
Checked 2 different devices the performance is not similar it is pretty solid. I still dont know why Voice Search capability check is so slow.